# Pandas: grouping

In [1]:
import pandas as pd
import numpy as np

In [2]:
cars = pd.read_csv("data/vehicles.csv")

In [3]:
cars.head()

Make                Model  Year  Engine Displacement  \
0        AM General    DJ Po Vehicle 2WD  1984                  2.5   
1        AM General     FJ8c Post Office  1984                  4.2   
2        AM General  Post Office DJ5 2WD  1985                  2.5   
3        AM General  Post Office DJ8 2WD  1985                  4.2   
4  ASC Incorporated                  GNX  1987                  3.8   

   Cylinders     Transmission        Drivetrain                Vehicle Class  \
0        4.0  Automatic 3-spd     2-Wheel Drive  Special Purpose Vehicle 2WD   
1        6.0  Automatic 3-spd     2-Wheel Drive  Special Purpose Vehicle 2WD   
2        4.0  Automatic 3-spd  Rear-Wheel Drive  Special Purpose Vehicle 2WD   
3        6.0  Automatic 3-spd  Rear-Wheel Drive  Special Purpose Vehicle 2WD   
4        6.0  Automatic 4-spd  Rear-Wheel Drive                 Midsize Cars   

  Fuel Type  Fuel Barrels/Year  City MPG  Highway MPG  Combined MPG  \
0   Regular          19.388824        18           17            17   
1   Regular          25.354615        13           13            13   
2   Regular          20.600625        16           17            16   
3   Regular          25.354615        13           13            13   
4   Premium          20.600625        14           21            16   

   CO2 Emission Grams/Mile  Fuel Cost/Year  
0               522.764706            1950  
1               683.615385            2550  
2               555.437500            2100  
3               683.615385            2550  
4               555.437500            2550

How many Car models? 

In [17]:
cars.Model.describe()


count               35952
unique               3608
top       F150 Pickup 2WD
freq                  197
Name: Model, dtype: object

*Answer: There are 3608 Car models.*

group by the data by the Make  using count function

In [11]:
model = cars['Model'].value_counts()
model

F150 Pickup 2WD      197
F150 Pickup 4WD      179
Truck 2WD            173
Mustang              170
Jetta                169
                    ... 
C300 Convertible       1
C300 Coupe             1
C300 FFV               1
C320 4matic Sedan      1
DJ Po Vehicle 2WD      1
Name: Model, Length: 3608, dtype: int64

What brand has the most cars?

*Answer: F150 Pickup 2WD*

<b>show the average CO2_Emission_Grams/Km  by Brand

In [22]:
cars.rename(columns = {"CO2 Emission Grams/Mile" : "CO2 Emission Grams/Km"}, inplace = True)

cars

Make                Model  Year  Engine Displacement  \
0            AM General    DJ Po Vehicle 2WD  1984                  2.5   
1            AM General     FJ8c Post Office  1984                  4.2   
2            AM General  Post Office DJ5 2WD  1985                  2.5   
3            AM General  Post Office DJ8 2WD  1985                  4.2   
4      ASC Incorporated                  GNX  1987                  3.8   
...                 ...                  ...   ...                  ...   
35947             smart         fortwo coupe  2013                  1.0   
35948             smart         fortwo coupe  2014                  1.0   
35949             smart         fortwo coupe  2015                  1.0   
35950             smart         fortwo coupe  2016                  0.9   
35951             smart         fortwo coupe  2016                  0.9   

       Cylinders     Transmission        Drivetrain  \
0            4.0  Automatic 3-spd     2-Wheel Drive   
1            6.0  Automatic 3-spd     2-Wheel Drive   
2            4.0  Automatic 3-spd  Rear-Wheel Drive   
3            6.0  Automatic 3-spd  Rear-Wheel Drive   
4            6.0  Automatic 4-spd  Rear-Wheel Drive   
...          ...              ...               ...   
35947        3.0        Auto(AM5)  Rear-Wheel Drive   
35948        3.0        Auto(AM5)  Rear-Wheel Drive   
35949        3.0        Auto(AM5)  Rear-Wheel Drive   
35950        3.0        Auto(AM6)  Rear-Wheel Drive   
35951        3.0     Manual 5-spd  Rear-Wheel Drive   

                     Vehicle Class Fuel Type  Fuel Barrels/Year  City MPG  \
0      Special Purpose Vehicle 2WD   Regular          19.388824        18   
1      Special Purpose Vehicle 2WD   Regular          25.354615        13   
2      Special Purpose Vehicle 2WD   Regular          20.600625        16   
3      Special Purpose Vehicle 2WD   Regular          25.354615        13   
4                     Midsize Cars   Premium          20.600625        14   
...                            ...       ...                ...       ...   
35947                  Two Seaters   Premium           9.155833        34   
35948                  Two Seaters   Premium           9.155833        34   
35949                  Two Seaters   Premium           9.155833        34   
35950                  Two Seaters   Premium           9.155833        34   
35951                  Two Seaters   Premium           9.417429        32   

       Highway MPG  Combined MPG  CO2 Emission Grams/Km  Fuel Cost/Year  
0               17            17             522.764706            1950  
1               13            13             683.615385            2550  
2               17            16             555.437500            2100  
3               13            13             683.615385            2550  
4               21            16             555.437500            2550  
...            ...           ...                    ...             ...  
35947           38            36             244.000000            1100  
35948           38            36             243.000000            1100  
35949           38            36             244.000000            1100  
35950           39            36             246.000000            1100  
35951           39            35             255.000000            1150  

[35952 rows x 15 columns]

In [23]:
cars["CO2 Emission Grams/Km"]=cars["CO2 Emission Grams/Km"] * (1/1.60934)
cars

Make                Model  Year  Engine Displacement  \
0            AM General    DJ Po Vehicle 2WD  1984                  2.5   
1            AM General     FJ8c Post Office  1984                  4.2   
2            AM General  Post Office DJ5 2WD  1985                  2.5   
3            AM General  Post Office DJ8 2WD  1985                  4.2   
4      ASC Incorporated                  GNX  1987                  3.8   
...                 ...                  ...   ...                  ...   
35947             smart         fortwo coupe  2013                  1.0   
35948             smart         fortwo coupe  2014                  1.0   
35949             smart         fortwo coupe  2015                  1.0   
35950             smart         fortwo coupe  2016                  0.9   
35951             smart         fortwo coupe  2016                  0.9   

       Cylinders     Transmission        Drivetrain  \
0            4.0  Automatic 3-spd     2-Wheel Drive   
1            6.0  Automatic 3-spd     2-Wheel Drive   
2            4.0  Automatic 3-spd  Rear-Wheel Drive   
3            6.0  Automatic 3-spd  Rear-Wheel Drive   
4            6.0  Automatic 4-spd  Rear-Wheel Drive   
...          ...              ...               ...   
35947        3.0        Auto(AM5)  Rear-Wheel Drive   
35948        3.0        Auto(AM5)  Rear-Wheel Drive   
35949        3.0        Auto(AM5)  Rear-Wheel Drive   
35950        3.0        Auto(AM6)  Rear-Wheel Drive   
35951        3.0     Manual 5-spd  Rear-Wheel Drive   

                     Vehicle Class Fuel Type  Fuel Barrels/Year  City MPG  \
0      Special Purpose Vehicle 2WD   Regular          19.388824        18   
1      Special Purpose Vehicle 2WD   Regular          25.354615        13   
2      Special Purpose Vehicle 2WD   Regular          20.600625        16   
3      Special Purpose Vehicle 2WD   Regular          25.354615        13   
4                     Midsize Cars   Premium          20.600625        14   
...                            ...       ...                ...       ...   
35947                  Two Seaters   Premium           9.155833        34   
35948                  Two Seaters   Premium           9.155833        34   
35949                  Two Seaters   Premium           9.155833        34   
35950                  Two Seaters   Premium           9.155833        34   
35951                  Two Seaters   Premium           9.417429        32   

       Highway MPG  Combined MPG  CO2 Emission Grams/Km  Fuel Cost/Year  
0               17            17             324.831736            1950  
1               13            13             424.779962            2550  
2               17            16             345.133719            2100  
3               13            13             424.779962            2550  
4               21            16             345.133719            2550  
...            ...           ...                    ...             ...  
35947           38            36             151.614948            1100  
35948           38            36             150.993575            1100  
35949           38            36             151.614948            1100  
35950           39            36             152.857693            1100  
35951           39            35             158.450048            1150  

[35952 rows x 15 columns]

In [24]:
cars.groupby("Make")["CO2 Emission Grams/Km"].apply(lambda grps:grps.mean())

Make
AM General                     379.881345
ASC Incorporated               345.133719
Acura                          262.583000
Alfa Romeo                     288.287195
American Motors Corporation    314.264744
                                  ...    
Volkswagen                     244.038998
Volvo                          270.796572
Wallace Environmental          408.857065
Yugo                           221.251107
smart                          153.498052
Name: CO2 Emission Grams/Km, Length: 127, dtype: float64

<b>show the average CO2_Emission_Grams/Km  by Brand ... sorted

In [26]:
print('cars.sort_values(by="CO2 Emission Grams/Km")')
cars.sort_values(by="CO2 Emission Grams/Km")


cars.sort_values(by="CO2 Emission Grams/Km")


Make     Model  Year  Engine Displacement  Cylinders  \
3071           BMW    i3 REX  2016                  0.6        2.0   
3069           BMW    i3 REX  2014                  0.6        2.0   
3070           BMW    i3 REX  2015                  0.6        2.0   
7916     Chevrolet      Volt  2016                  1.5        4.0   
7917     Chevrolet      Volt  2017                  1.5        4.0   
...            ...       ...   ...                  ...        ...   
20897  Lamborghini  Countach  1989                  5.2       12.0   
20894  Lamborghini  Countach  1986                  5.2       12.0   
20898  Lamborghini  Countach  1990                  5.2       12.0   
20896  Lamborghini  Countach  1988                  5.2       12.0   
20895  Lamborghini  Countach  1987                  5.2       12.0   

                           Transmission         Drivetrain    Vehicle Class  \
3071                     Automatic (A1)   Rear-Wheel Drive  Subcompact Cars   
3069                           Auto(A1)   Rear-Wheel Drive  Subcompact Cars   
3070                     Automatic (A1)   Rear-Wheel Drive  Subcompact Cars   
7916   Automatic (variable gear ratios)  Front-Wheel Drive     Compact Cars   
7917   Automatic (variable gear ratios)  Front-Wheel Drive     Compact Cars   
...                                 ...                ...              ...   
20897                      Manual 5-spd   Rear-Wheel Drive      Two Seaters   
20894                      Manual 5-spd   Rear-Wheel Drive      Two Seaters   
20898                      Manual 5-spd   Rear-Wheel Drive      Two Seaters   
20896                      Manual 5-spd   Rear-Wheel Drive      Two Seaters   
20895                      Manual 5-spd   Rear-Wheel Drive      Two Seaters   

                        Fuel Type  Fuel Barrels/Year  City MPG  Highway MPG  \
3071   Premium Gas or Electricity           1.563190        41           37   
3069   Premium Gas or Electricity           1.563190        41           37   
3070   Premium Gas or Electricity           1.563190        41           37   
7916   Regular Gas or Electricity           2.006844        43           42   
7917   Regular Gas or Electricity           2.006844        43           42   
...                           ...                ...       ...          ...   
20897                     Premium          47.087143         6           10   
20894                     Premium          47.087143         6           10   
20898                     Premium          47.087143         6           10   
20896                     Premium          47.087143         6           10   
20895                     Premium          47.087143         6           10   

       Combined MPG  CO2 Emission Grams/Km  Fuel Cost/Year  
3071             39              22.990791            1050  
3069             39              24.854909            1050  
3070             39              24.854909            1050  
7916             42              31.690010             800  
7917             42              31.690010             800  
...             ...                    ...             ...  
20897             7             788.877073            5800  
20894             7             788.877073            5800  
20898             7             788.877073            5800  
20896             7             788.877073            5800  
20895             7             788.877073            5800  

[35952 rows x 15 columns]

Use `pd.cut` or `pd.qcut` to create 4 groups (bins) of cars, by Year. We want to explore how cars have evolved decade by decade.

In [27]:
cars['Year'].describe()

count    35952.00000
mean      2000.71640
std         10.08529
min       1984.00000
25%       1991.00000
50%       2001.00000
75%       2010.00000
max       2017.00000
Name: Year, dtype: float64

In [30]:
#perguntar Nelson

pd.cut(cars['Year'],4)

0        (1983.967, 1992.25]
1        (1983.967, 1992.25]
2        (1983.967, 1992.25]
3        (1983.967, 1992.25]
4        (1983.967, 1992.25]
                ...         
35947      (2008.75, 2017.0]
35948      (2008.75, 2017.0]
35949      (2008.75, 2017.0]
35950      (2008.75, 2017.0]
35951      (2008.75, 2017.0]
Name: Year, Length: 35952, dtype: category
Categories (4, interval[float64, right]): [(1983.967, 1992.25] < (1992.25, 2000.5] < (2000.5, 2008.75] < (2008.75, 2017.0]]

### Did cars consume more gas in the eighties?

show the average City_Km/Liter by year_range

In [ ]:
#Converter City Miles por galão (3,6l)

In [34]:
cars.rename(columns = {"City MPG" : "City_Km/Liter"}, inplace = True)

cars

Make                Model  Year  Engine Displacement  \
0            AM General    DJ Po Vehicle 2WD  1984                  2.5   
1            AM General     FJ8c Post Office  1984                  4.2   
2            AM General  Post Office DJ5 2WD  1985                  2.5   
3            AM General  Post Office DJ8 2WD  1985                  4.2   
4      ASC Incorporated                  GNX  1987                  3.8   
...                 ...                  ...   ...                  ...   
35947             smart         fortwo coupe  2013                  1.0   
35948             smart         fortwo coupe  2014                  1.0   
35949             smart         fortwo coupe  2015                  1.0   
35950             smart         fortwo coupe  2016                  0.9   
35951             smart         fortwo coupe  2016                  0.9   

       Cylinders     Transmission        Drivetrain  \
0            4.0  Automatic 3-spd     2-Wheel Drive   
1            6.0  Automatic 3-spd     2-Wheel Drive   
2            4.0  Automatic 3-spd  Rear-Wheel Drive   
3            6.0  Automatic 3-spd  Rear-Wheel Drive   
4            6.0  Automatic 4-spd  Rear-Wheel Drive   
...          ...              ...               ...   
35947        3.0        Auto(AM5)  Rear-Wheel Drive   
35948        3.0        Auto(AM5)  Rear-Wheel Drive   
35949        3.0        Auto(AM5)  Rear-Wheel Drive   
35950        3.0        Auto(AM6)  Rear-Wheel Drive   
35951        3.0     Manual 5-spd  Rear-Wheel Drive   

                     Vehicle Class Fuel Type  Fuel Barrels/Year  \
0      Special Purpose Vehicle 2WD   Regular          19.388824   
1      Special Purpose Vehicle 2WD   Regular          25.354615   
2      Special Purpose Vehicle 2WD   Regular          20.600625   
3      Special Purpose Vehicle 2WD   Regular          25.354615   
4                     Midsize Cars   Premium          20.600625   
...                            ...       ...                ...   
35947                  Two Seaters   Premium           9.155833   
35948                  Two Seaters   Premium           9.155833   
35949                  Two Seaters   Premium           9.155833   
35950                  Two Seaters   Premium           9.155833   
35951                  Two Seaters   Premium           9.417429   

       City_Km/Liter  Highway MPG  Combined MPG  CO2 Emission Grams/Km  \
0                 18           17            17             324.831736   
1                 13           13            13             424.779962   
2                 16           17            16             345.133719   
3                 13           13            13             424.779962   
4                 14           21            16             345.133719   
...              ...          ...           ...                    ...   
35947             34           38            36             151.614948   
35948             34           38            36             150.993575   
35949             34           38            36             151.614948   
35950             34           39            36             152.857693   
35951             32           39            35             158.450048   

       Fuel Cost/Year  
0                1950  
1                2550  
2                2100  
3                2550  
4                2550  
...               ...  
35947            1100  
35948            1100  
35949            1100  
35950            1100  
35951            1150  

[35952 rows x 15 columns]

In [36]:
cars["City_Km/Liter"]=cars["City_Km/Liter"] * (2.35215)
cars

Make                Model  Year  Engine Displacement  \
0            AM General    DJ Po Vehicle 2WD  1984                  2.5   
1            AM General     FJ8c Post Office  1984                  4.2   
2            AM General  Post Office DJ5 2WD  1985                  2.5   
3            AM General  Post Office DJ8 2WD  1985                  4.2   
4      ASC Incorporated                  GNX  1987                  3.8   
...                 ...                  ...   ...                  ...   
35947             smart         fortwo coupe  2013                  1.0   
35948             smart         fortwo coupe  2014                  1.0   
35949             smart         fortwo coupe  2015                  1.0   
35950             smart         fortwo coupe  2016                  0.9   
35951             smart         fortwo coupe  2016                  0.9   

       Cylinders     Transmission        Drivetrain  \
0            4.0  Automatic 3-spd     2-Wheel Drive   
1            6.0  Automatic 3-spd     2-Wheel Drive   
2            4.0  Automatic 3-spd  Rear-Wheel Drive   
3            6.0  Automatic 3-spd  Rear-Wheel Drive   
4            6.0  Automatic 4-spd  Rear-Wheel Drive   
...          ...              ...               ...   
35947        3.0        Auto(AM5)  Rear-Wheel Drive   
35948        3.0        Auto(AM5)  Rear-Wheel Drive   
35949        3.0        Auto(AM5)  Rear-Wheel Drive   
35950        3.0        Auto(AM6)  Rear-Wheel Drive   
35951        3.0     Manual 5-spd  Rear-Wheel Drive   

                     Vehicle Class Fuel Type  Fuel Barrels/Year  \
0      Special Purpose Vehicle 2WD   Regular          19.388824   
1      Special Purpose Vehicle 2WD   Regular          25.354615   
2      Special Purpose Vehicle 2WD   Regular          20.600625   
3      Special Purpose Vehicle 2WD   Regular          25.354615   
4                     Midsize Cars   Premium          20.600625   
...                            ...       ...                ...   
35947                  Two Seaters   Premium           9.155833   
35948                  Two Seaters   Premium           9.155833   
35949                  Two Seaters   Premium           9.155833   
35950                  Two Seaters   Premium           9.155833   
35951                  Two Seaters   Premium           9.417429   

       City_Km/Liter  Highway MPG  Combined MPG  CO2 Emission Grams/Km  \
0           42.33870           17            17             324.831736   
1           30.57795           13            13             424.779962   
2           37.63440           17            16             345.133719   
3           30.57795           13            13             424.779962   
4           32.93010           21            16             345.133719   
...              ...          ...           ...                    ...   
35947       79.97310           38            36             151.614948   
35948       79.97310           38            36             150.993575   
35949       79.97310           38            36             151.614948   
35950       79.97310           39            36             152.857693   
35951       75.26880           39            35             158.450048   

       Fuel Cost/Year  
0                1950  
1                2550  
2                2100  
3                2550  
4                2550  
...               ...  
35947            1100  
35948            1100  
35949            1100  
35950            1100  
35951            1150  

[35952 rows x 15 columns]

In [55]:
cars['Decade'] = pd.cut(cars['Year'],bins=4,labels=['80s','90s','00s','10s'])
cars[['Year','Decade']]

Year Decade
0      1984    80s
1      1984    80s
2      1985    80s
3      1985    80s
4      1987    80s
...     ...    ...
35947  2013    10s
35948  2014    10s
35949  2015    10s
35950  2016    10s
35951  2016    10s

[35952 rows x 2 columns]

In [58]:
cars.groupby("Decade")["City_Km/Liter"].apply(lambda grps:grps.mean())


Decade
80s    40.421773
90s    39.953335
00s    39.404204
10s    45.378972
Name: City_Km/Liter, dtype: float64

In [59]:
cars.groupby(["Decade"])[['Decade',"City_Km/Liter"]].mean("City_Km/Liter")  

City_Km/Liter
Decade               
80s         40.421773
90s         39.953335
00s         39.404204
10s         45.378972

*Answer: Yes. Apparently cars are consuming more fuel over the years.*

Which brands are more environment friendly?

In [33]:
cars.groupby("Make")["CO2 Emission Grams/Km"].apply(lambda grps:grps.mean())

Make
AM General                     379.881345
ASC Incorporated               345.133719
Acura                          262.583000
Alfa Romeo                     288.287195
American Motors Corporation    314.264744
                                  ...    
Volkswagen                     244.038998
Volvo                          270.796572
Wallace Environmental          408.857065
Yugo                           221.251107
smart                          153.498052
Name: CO2 Emission Grams/Km, Length: 127, dtype: float64

*Answer: Smart, Yugo and Wallace Environmental*

Does the drivetrain affect fuel consumption?

In [38]:
##Converting Highway MPG to Highway_Km/Liter
cars.rename(columns = {"Highway MPG" : "Highway_Km/Liter"}, inplace = True)

cars

Make                Model  Year  Engine Displacement  \
0            AM General    DJ Po Vehicle 2WD  1984                  2.5   
1            AM General     FJ8c Post Office  1984                  4.2   
2            AM General  Post Office DJ5 2WD  1985                  2.5   
3            AM General  Post Office DJ8 2WD  1985                  4.2   
4      ASC Incorporated                  GNX  1987                  3.8   
...                 ...                  ...   ...                  ...   
35947             smart         fortwo coupe  2013                  1.0   
35948             smart         fortwo coupe  2014                  1.0   
35949             smart         fortwo coupe  2015                  1.0   
35950             smart         fortwo coupe  2016                  0.9   
35951             smart         fortwo coupe  2016                  0.9   

       Cylinders     Transmission        Drivetrain  \
0            4.0  Automatic 3-spd     2-Wheel Drive   
1            6.0  Automatic 3-spd     2-Wheel Drive   
2            4.0  Automatic 3-spd  Rear-Wheel Drive   
3            6.0  Automatic 3-spd  Rear-Wheel Drive   
4            6.0  Automatic 4-spd  Rear-Wheel Drive   
...          ...              ...               ...   
35947        3.0        Auto(AM5)  Rear-Wheel Drive   
35948        3.0        Auto(AM5)  Rear-Wheel Drive   
35949        3.0        Auto(AM5)  Rear-Wheel Drive   
35950        3.0        Auto(AM6)  Rear-Wheel Drive   
35951        3.0     Manual 5-spd  Rear-Wheel Drive   

                     Vehicle Class Fuel Type  Fuel Barrels/Year  \
0      Special Purpose Vehicle 2WD   Regular          19.388824   
1      Special Purpose Vehicle 2WD   Regular          25.354615   
2      Special Purpose Vehicle 2WD   Regular          20.600625   
3      Special Purpose Vehicle 2WD   Regular          25.354615   
4                     Midsize Cars   Premium          20.600625   
...                            ...       ...                ...   
35947                  Two Seaters   Premium           9.155833   
35948                  Two Seaters   Premium           9.155833   
35949                  Two Seaters   Premium           9.155833   
35950                  Two Seaters   Premium           9.155833   
35951                  Two Seaters   Premium           9.417429   

       City_Km/Liter  Highway_Km/Liter  Combined MPG  CO2 Emission Grams/Km  \
0           42.33870                17            17             324.831736   
1           30.57795                13            13             424.779962   
2           37.63440                17            16             345.133719   
3           30.57795                13            13             424.779962   
4           32.93010                21            16             345.133719   
...              ...               ...           ...                    ...   
35947       79.97310                38            36             151.614948   
35948       79.97310                38            36             150.993575   
35949       79.97310                38            36             151.614948   
35950       79.97310                39            36             152.857693   
35951       75.26880                39            35             158.450048   

       Fuel Cost/Year  
0                1950  
1                2550  
2                2100  
3                2550  
4                2550  
...               ...  
35947            1100  
35948            1100  
35949            1100  
35950            1100  
35951            1150  

[35952 rows x 15 columns]

In [39]:
cars["Highway_Km/Liter"]=cars["Highway_Km/Liter"] * (2.35215)
cars

Make                Model  Year  Engine Displacement  \
0            AM General    DJ Po Vehicle 2WD  1984                  2.5   
1            AM General     FJ8c Post Office  1984                  4.2   
2            AM General  Post Office DJ5 2WD  1985                  2.5   
3            AM General  Post Office DJ8 2WD  1985                  4.2   
4      ASC Incorporated                  GNX  1987                  3.8   
...                 ...                  ...   ...                  ...   
35947             smart         fortwo coupe  2013                  1.0   
35948             smart         fortwo coupe  2014                  1.0   
35949             smart         fortwo coupe  2015                  1.0   
35950             smart         fortwo coupe  2016                  0.9   
35951             smart         fortwo coupe  2016                  0.9   

       Cylinders     Transmission        Drivetrain  \
0            4.0  Automatic 3-spd     2-Wheel Drive   
1            6.0  Automatic 3-spd     2-Wheel Drive   
2            4.0  Automatic 3-spd  Rear-Wheel Drive   
3            6.0  Automatic 3-spd  Rear-Wheel Drive   
4            6.0  Automatic 4-spd  Rear-Wheel Drive   
...          ...              ...               ...   
35947        3.0        Auto(AM5)  Rear-Wheel Drive   
35948        3.0        Auto(AM5)  Rear-Wheel Drive   
35949        3.0        Auto(AM5)  Rear-Wheel Drive   
35950        3.0        Auto(AM6)  Rear-Wheel Drive   
35951        3.0     Manual 5-spd  Rear-Wheel Drive   

                     Vehicle Class Fuel Type  Fuel Barrels/Year  \
0      Special Purpose Vehicle 2WD   Regular          19.388824   
1      Special Purpose Vehicle 2WD   Regular          25.354615   
2      Special Purpose Vehicle 2WD   Regular          20.600625   
3      Special Purpose Vehicle 2WD   Regular          25.354615   
4                     Midsize Cars   Premium          20.600625   
...                            ...       ...                ...   
35947                  Two Seaters   Premium           9.155833   
35948                  Two Seaters   Premium           9.155833   
35949                  Two Seaters   Premium           9.155833   
35950                  Two Seaters   Premium           9.155833   
35951                  Two Seaters   Premium           9.417429   

       City_Km/Liter  Highway_Km/Liter  Combined MPG  CO2 Emission Grams/Km  \
0           42.33870          39.98655            17             324.831736   
1           30.57795          30.57795            13             424.779962   
2           37.63440          39.98655            16             345.133719   
3           30.57795          30.57795            13             424.779962   
4           32.93010          49.39515            16             345.133719   
...              ...               ...           ...                    ...   
35947       79.97310          89.38170            36             151.614948   
35948       79.97310          89.38170            36             150.993575   
35949       79.97310          89.38170            36             151.614948   
35950       79.97310          91.73385            36             152.857693   
35951       75.26880          91.73385            35             158.450048   

       Fuel Cost/Year  
0                1950  
1                2550  
2                2100  
3                2550  
4                2550  
...               ...  
35947            1100  
35948            1100  
35949            1100  
35950            1100  
35951            1150  

[35952 rows x 15 columns]

In [49]:
#the values are different from the example

cars.groupby("Drivetrain")["Highway_Km/Liter","City_Km/Liter"].apply(lambda grps:grps.mean())

/var/folders/fz/8c6p2kfn5z98lys6tff8f9jm0000gn/T/ipykernel_15715/1431936493.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  cars.groupby("Drivetrain")["Highway_Km/Liter","City_Km/Liter"].apply(lambda grps:grps.mean())


Highway_Km/Liter  City_Km/Liter
Drivetrain                                                 
2-Wheel Drive                      45.491582      36.750259
2-Wheel Drive, Front               77.620950      58.803750
4-Wheel Drive                      53.492515      39.784238
4-Wheel or All-Wheel Drive         46.181425      35.364718
All-Wheel Drive                    60.208811      43.074684
Front-Wheel Drive                  67.310909      49.805748
Part-time 4-Wheel Drive            44.899268      34.389028
Rear-Wheel Drive                   49.925983      36.274973

*Answer: Yes. Apparently 2-Wheel Drive consumes more fuel.*

Do cars with automatic transmission consume more fuel than cars with manual transmission?

In [65]:
cars['Transmission'].unique()

array(['Automatic 3-spd', 'Automatic 4-spd', 'Manual 5-spd',
       'Automatic (S5)', 'Manual 6-spd', 'Automatic 5-spd', 'Auto(AM8)',
       'Auto(AM-S8)', 'Auto(AV-S7)', 'Automatic (S6)', 'Automatic (S9)',
       'Automatic (S4)', 'Auto(AM-S9)', 'Automatic (S7)', 'Auto(AM7)',
       'Auto(AM-S7)', 'Auto(AM6)', 'Automatic 6-spd', 'Manual 4-spd',
       'Automatic (S8)', 'Manual(M7)', 'Auto(AM-S6)',
       'Automatic (variable gear ratios)', 'Automatic (AV)',
       'Auto(AV-S8)', 'Automatic (AM6)', 'Automatic 8-spd', 'Auto(A1)',
       'Automatic (A1)', 'Automatic (A6)', 'Auto(AV-S6)', 'Manual 3-spd',
       'Manual 7-spd', 'Automatic 9-spd', 'Auto (AV)', 'Automatic 6spd',
       'Auto(L4)', 'Auto(L3)', 'Auto (AV-S6)', 'Auto (AV-S8)',
       'Automatic (AV-S6)', 'Automatic 7-spd', 'Manual 5 spd',
       'Auto(AM5)', 'Automatic (AM5)'], dtype=object)

In [67]:
cars['Transmission'] = list( map(lambda x: "Automatic" if ("Auto" in x) else "Manual",cars['Transmission']) )
cars

Make                Model  Year  Engine Displacement  \
0            AM General    DJ Po Vehicle 2WD  1984                  2.5   
1            AM General     FJ8c Post Office  1984                  4.2   
2            AM General  Post Office DJ5 2WD  1985                  2.5   
3            AM General  Post Office DJ8 2WD  1985                  4.2   
4      ASC Incorporated                  GNX  1987                  3.8   
...                 ...                  ...   ...                  ...   
35947             smart         fortwo coupe  2013                  1.0   
35948             smart         fortwo coupe  2014                  1.0   
35949             smart         fortwo coupe  2015                  1.0   
35950             smart         fortwo coupe  2016                  0.9   
35951             smart         fortwo coupe  2016                  0.9   

       Cylinders Transmission        Drivetrain                Vehicle Class  \
0            4.0    Automatic     2-Wheel Drive  Special Purpose Vehicle 2WD   
1            6.0    Automatic     2-Wheel Drive  Special Purpose Vehicle 2WD   
2            4.0    Automatic  Rear-Wheel Drive  Special Purpose Vehicle 2WD   
3            6.0    Automatic  Rear-Wheel Drive  Special Purpose Vehicle 2WD   
4            6.0    Automatic  Rear-Wheel Drive                 Midsize Cars   
...          ...          ...               ...                          ...   
35947        3.0    Automatic  Rear-Wheel Drive                  Two Seaters   
35948        3.0    Automatic  Rear-Wheel Drive                  Two Seaters   
35949        3.0    Automatic  Rear-Wheel Drive                  Two Seaters   
35950        3.0    Automatic  Rear-Wheel Drive                  Two Seaters   
35951        3.0       Manual  Rear-Wheel Drive                  Two Seaters   

      Fuel Type  Fuel Barrels/Year  City_Km/Liter  Highway_Km/Liter  \
0       Regular          19.388824       42.33870          39.98655   
1       Regular          25.354615       30.57795          30.57795   
2       Regular          20.600625       37.63440          39.98655   
3       Regular          25.354615       30.57795          30.57795   
4       Premium          20.600625       32.93010          49.39515   
...         ...                ...            ...               ...   
35947   Premium           9.155833       79.97310          89.38170   
35948   Premium           9.155833       79.97310          89.38170   
35949   Premium           9.155833       79.97310          89.38170   
35950   Premium           9.155833       79.97310          91.73385   
35951   Premium           9.417429       75.26880          91.73385   

       Combined MPG  CO2 Emission Grams/Km  Fuel Cost/Year Decade  
0                17             324.831736            1950    80s  
1                13             424.779962            2550    80s  
2                16             345.133719            2100    80s  
3                13             424.779962            2550    80s  
4                16             345.133719            2550    80s  
...             ...                    ...             ...    ...  
35947            36             151.614948            1100    10s  
35948            36             150.993575            1100    10s  
35949            36             151.614948            1100    10s  
35950            36             152.857693            1100    10s  
35951            35             158.450048            1150    10s  

[35952 rows x 16 columns]

In [68]:
consumption = cars.groupby('Transmission')['City_Km/Liter'].count()
consumption

Transmission
Automatic    24290
Manual       11662
Name: City_Km/Liter, dtype: int64

In [69]:
cars.groupby("Transmission")[["City_Km/Liter"]].mean().sort_values("City_Km/Liter",ascending=False)

City_Km/Liter
Transmission               
Manual            44.085769
Automatic         40.267956

Use `groupby` and `aggregate` with different aggregation measures for different columns:

aggregate with average City_Km/Liter and the count of the Trans

In [143]:
## your code is here

City_Km/Liter  Trans
Trans                          
Automatic       7.278292  24290
Manual          7.968348  11662

In [70]:
cars.groupby("Transmission").aggregate({"City_Km/Liter":"mean","Transmission":"count"})

City_Km/Liter  Transmission
Transmission                             
Automatic         40.267956         24290
Manual            44.085769         11662

aggregate with average City_Km/Liter and the minimum of the Trans

In [144]:
### your code is here

City_Km/Liter
Trans                   
Automatic       2.976000
Manual          2.550857

In [71]:
cars.groupby("Transmission").aggregate({"City_Km/Liter":"mean","City_Km/Liter":"min"})

City_Km/Liter
Transmission               
Automatic          16.46505
Manual             14.11290